src/hook.rs

```diff
@@ -0,0 +1,175 @@
+    /// safety layer and convenience wrapped over Lua's low-level hooks. `1` is such an example of
+    /// a high overhead choice.
+    ///
+    /// Please find a number that is high enough so it's not that bad of an issue, while still
+    /// having enough precision for your needs. Keep in mind instructions are additions, calls to
+    /// functions, assignments to variables, etc.; they are very short.
+    pub every_nth_instruction: Option<u32>,
+}
+
+impl HookTriggers {
+    /// Compute the mask to pass to `lua_sethook`.
+    pub(crate) fn mask(&self) -> c_int {
+        let mut mask: c_int = 0;
+        if self.on_calls { mask |= ffi::LUA_MASKCALL }
+        if self.on_returns { mask |= ffi::LUA_MASKRET }
+        if self.every_line { mask |= ffi::LUA_MASKLINE }
+        if self.every_nth_instruction.is_some() { mask |= ffi::LUA_MASKCOUNT }
+        mask
+    }
+
+    /// Returns the `count` parameter to pass to `lua_sethook`, if applicable. Otherwise, zero is
+    /// returned.
+    pub(crate) fn count(&self) -> c_int {
+        self.every_nth_instruction.unwrap_or(0) as c_int
+    }
+}
+
+impl Default for HookTriggers {
+    fn default() -> Self {
+        HookTriggers {
+            on_calls: false,
+            on_returns: false,
+            every_line: false,
+            every_nth_instruction: None
+        }
+    }
+}
+
+/// This callback is passed to `lua_sethook` and gets called whenever debug information is received.
+pub(crate) unsafe extern "C" fn hook_proc(state: *mut lua_State, ar: *mut lua_Debug) {
+    callback_error(state, || {
+        if ffi::lua_getinfo(state, cstr!("nSltu"), ar) == 0 {
+            rlua_panic!("lua_getinfo failed")
+        }
+
+        let lua = Lua::from_state(state);
+        let debug = Debug {
+            name: ptr_to_str((*ar).name as *const i8),
+            namewhat: ptr_to_str((*ar).namewhat as *const i8),
+            what: ptr_to_str((*ar).what as *const i8),
+            source: ptr_to_str((*ar).source as *const i8),
+            curr_line: (*ar).currentline as u32,
+            line_defined: (*ar).linedefined as u32,
+            last_line_defined: (*ar).lastlinedefined as u32,
+            num_ups: (*ar).nups as u32,
+            num_params: (*ar).nparams as u32,
+            is_vararg: (*ar).isvararg == 1,
+            is_tailcall: (*ar).istailcall == 1,
+            short_src: str::from_utf8(mem::transmute((*ar).short_src.as_ref()))
+                .and_then(|r| Ok(Some(Cow::from(r))))
+                .unwrap_or(None),
+            _unused: ()
+        };
+
+        let cb = (&*extra_data(state)).hook_callback
+            .as_ref()
+            .expect("rlua internal error: no hooks previously set; this is a bug");
+        (&mut *match cb.try_borrow_mut() {
+            Ok(b) => b,
+            Err(_) => rlua_panic!("Lua should not allow hooks to be called within another hook;\
+                please make an issue")
+        })(lua, &debug)
```

Comment:
This is still UB, because you're not cloning the Rc you're borrowing it.

Reply:
Forgot about that!